src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,411 @@
+	{true, true, true, VersionTLS13, VersionTLS13, true, true, "tls13, using server skip verify"},
+	{false, true, false, VersionTLS13, VersionTLS13, true, false, "client no DC"},
+	{true, false, false, VersionTLS13, VersionTLS13, true, false, "server no DC"},
+	{true, false, false, VersionTLS12, VersionTLS13, true, false, "client using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS12, true, false, "server using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS11, VersionTLS13, true, false, "client using TLS 1.1. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS10, false, false, "server using TLS 1.0. No DC is supported in that version."},
+}
+
+// Checks that the client suppports a version >= 1.3 and accepts delegated
+// credentials. If so, it returns the delegation certificate; otherwise it
+// returns a non-delegated certificate.
+func testServerGetCertificate(ch *ClientHelloInfo) (*Certificate, error) {
+	versOk := false
+	for _, vers := range ch.SupportedVersions {
+		versOk = versOk || (vers >= uint16(VersionTLS13))
+	}
+
+	if versOk && ch.SupportDelegatedCredential {
+		return dcTestCerts["dc"], nil
+	}
+	return dcTestCerts["no dc"], nil
+
+}
+
+// Checks that the client supports the signature algorithm supported by the test
+// server, and that the server has a Delegated Credential.
+func testServerGetDelegatedCredential(ch *ClientHelloInfo) (*DelegatedCredential, crypto.PrivateKey, error) {
+	schemeOk := false
+	for _, scheme := range ch.SignatureSchemes {
+		schemeOk = schemeOk || (scheme == dcTestDCScheme)
+	}
+
+	if schemeOk && ch.SupportDelegatedCredential {
+		return dcTest.DelegatedCredential, dcTest.privateKey, nil
+	}
+	return nil, nil, nil
+}
+
+// Tests the handshake and one round of application data. Returns true if the
+// connection correctly used a Delegated Credential.
+func testConnWithDC(t *testing.T, clientMsg, serverMsg string, clientConfig, serverConfig *Config) (bool, error) {
+	initialize()
+	ln := newLocalListener(t)
+	defer ln.Close()
+
+	serverCh := make(chan *Conn, 1)
+	var serverErr error
+	go func() {
+		serverConn, err := ln.Accept()
+		if err != nil {
+			serverErr = err
+			serverCh <- nil
+			return
+		}
+		server := Server(serverConn, serverConfig)
+		if err := server.Handshake(); err != nil {
+			serverErr = fmt.Errorf("handshake error: %v", err)
+			serverCh <- nil
+			return
+		}
+		serverCh <- server
+	}()
+
+	client, err := Dial("tcp", ln.Addr().String(), clientConfig)
+	if err != nil {
+		return false, err
+	}
+	defer client.Close()
+
+	server := <-serverCh
+	if server == nil {
+		return false, serverErr
+	}
+
+	bufLen := len(clientMsg)
+	if len(serverMsg) > len(clientMsg) {
+		bufLen = len(serverMsg)
+	}
+	buf := make([]byte, bufLen)
+
+	client.Write([]byte(clientMsg))
+	n, err := server.Read(buf)
+	if n != len(clientMsg) || string(buf[:n]) != clientMsg {
```

Comment:
You probably want ``|| err`` here, like below?

Reply:
Added now.